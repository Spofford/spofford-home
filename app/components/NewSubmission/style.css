.submission-form-container {
  width:400px;
  margin:0px auto;
}

.wrapper {
  display: flex;
  flex-flow: column nowrap;
  justify-content: center;
  padding-bottom: 2rem;
}

.form {
  display: flex;
  flex-direction: column;
  justify-content: center;
  padding: 2rem;
}

.input, .half-input {
  padding: 1rem 1rem;
  border-radius: 3px;
  border: 1px solid #ccc;
  font-size: 1.1em;
  outline: none;
}

.inputGroup {
  display: flex;
  flex-flow: column nowrap;
  padding: 10px 0;
}

.new-submission h2 {
  margin: 20px 0px;
}

#dropzone {
  font-size:1.2em;
}

textarea {
  width:100%;
}

.new-submission .submission-form-container > button {
  float:right;
}

.submission-form-container {
  margin-bottom:40px;
}

.new-submission a {
  margin-top: 3px;
  display:block;
}

.__react_component_tooltip {
  max-width:400px !important;
}
