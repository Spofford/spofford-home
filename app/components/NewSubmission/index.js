import React from 'react'
import ReactDOM from 'react-dom'
import ReactCrop, { makeAspectCrop } from 'react-image-crop'
import 'react-image-crop/dist/ReactCrop.css'
import cssModules from "react-css-modules"
import style from "./style.css"
import { connect } from "react-redux"
import Actions from "../../redux/actions"
import Dropzone from 'react-dropzone'
import 'react-image-crop/dist/ReactCrop.css';
import {image64toCanvasRef, extractImageFileExtensionFromBase64} from '../../../src/ReusableUtils.js'
import { default as Header } from "../Header"
import classNames from 'classnames';
import { Redirect } from 'react-router-dom'

const imageMaxSize = 10000000

export class NewSubmission extends React.Component {
  constructor(props){
    super(props)
    this.imagePreviewCanvasRef = React.createRef()
    this.state = {
      src: null,
      completedSrc: null,
      crop: {
        aspect: 1
      },
      photoStart: true,
      photoCrop: false,
      photoDone: false,
      redirect: false
    }
    this.submit = this.submit.bind(this)
    // this.handleCompleteClick = this.submit.bind(this)
  }

  submit() {
    const submission = {
      description: document.getElementById("submission-description").value,
      manufacturing: document.getElementById("submission-manufacturing").value,
      cad_url: document.getElementById("submission-cad").value,
      user_id: this.props.user.id
    }
    const image = this.state.completedSrc.split(',')[1]
    this.props.dispatch(Actions.submissionNew(submission, image))
    //this.setState({redirect: true})
  }

  handleOnDrop = (files, rejectedFiles) => {
    if (rejectedFiles && rejectedFiles.length > 0) {
      const currentRejectFile = rejectedFiles[0]
      const currentRejectFileType = currentRejectFile.type
      const currentRejectFileSize = currentRejectFile.size
      if(currentRejectFileSize > imageMaxSize) {
        alert("this file is too big")
      }
    }

    if (files && files.length > 0) {
      const currentFile = files[0]
      const currentFileType = currentFile.type
      const currentFileSize = currentFile.size
      if(currentFileSize > imageMaxSize) {
        alert("this file is too big")
      }
    }

    const reader = new FileReader()
    reader.addEventListener(
      'load',
      () =>
        this.setState({
          src: reader.result,
        }),
      false
    )
    reader.readAsDataURL(files[0])

    this.setState({
      photoStart: false,
      photoCrop: true,
      photoDone: false
    })
  }

  onImageLoaded = image => {

  }

  onCropComplete = (crop, pixelCrop) => {

    const canvasRef = this.imagePreviewCanvasRef.current
    const {src} = this.state
    image64toCanvasRef(canvasRef, src, pixelCrop)
  }

  handleCompleteClick = (event) => {
    event.preventDefault()
    const canvasRef = this.imagePreviewCanvasRef.current
    const {src} = this.state
    const fileExtension = extractImageFileExtensionFromBase64(src)
    const imageData64 = canvasRef.toDataURL('image/' + fileExtension)

    this.setState({
      completedSrc: imageData64,
      photoStart: false,
      photoCrop: false,
      photoDone: true
    })
  }

  onCropChange = crop => {
    this.setState({ crop })
  }

  render() {
    var startClasses = classNames(
      {'phase_one': this.state.photoStart},
      {'phase_two': this.state.photoCrop},
      {'phase_three': this.state.photoDone}
    )

    const { redirect } = this.state;


     if(this.props.submission.description) {
       return <Redirect to='/submissions'/>;
     }

    return (
      <div className="submissions">
        <Header />
        <div className="body-container">
          <div className="submission-form-container">
            <h2>New Submission</h2>
            <Dropzone className={startClasses} id="dropzone" onDrop={this.handleOnDrop} maxSize={imageMaxSize} multiple={false}>Drop photo here or click anywhere to choose a file (.png and .jpg files only, 10MB maximum)</Dropzone>
            {this.state.src && (
              <ReactCrop
                className={startClasses}
                src={this.state.src}
                crop={this.state.crop}
                onImageLoaded={this.onImageLoaded}
                onComplete={this.onCropComplete}
                onChange={this.onCropChange}
              />
            )}
            <canvas className={startClasses} width="400" height="400" ref={this.imagePreviewCanvasRef}></canvas>
            <button className={startClasses} id="crop-button" onClick={this.handleCompleteClick}>Click Me</button>
            <div className="inputGroup">
              <textarea
                placeholder="Description"
                className="input"
                type="text"
                autoComplete="description"
                id="submission-description" />
            </div>
            <div className="inputGroup">
              <textarea
                placeholder="Manufacturing Description"
                className="input"
                type="text"
                autoComplete="manufacturing"
                id="submission-manufacturing" />
            </div>
            <div className="inputGroup">
              <input
                placeholder="CAD file"
                className="input"
                type="text"
                autoComplete="cad-file"
                id="submission-cad" />
            </div>
            <button onClick={this.submit}>Submit</button>
          </div>
        </div>
      </div>
    )
  }
}

const mapStateToProps = state => ({
  user: state.user,
  submission: state.submission
})

export default connect(mapStateToProps)(cssModules(NewSubmission, style))
