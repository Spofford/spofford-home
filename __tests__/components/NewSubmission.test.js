import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';

import NewSubmission from '../../app/components/NewSubmission';

const mockStore = configureStore();
const initialState = {
  user: {
    email: "amhasler@gmail.com",
    first_name: "Adam",
    last_name: "Hasler",
    role: 1,
    id: 20,
    charges: []
  },
  submission: {
    description: "description one",
    manufacturing: "manufacturing description one",
    advance: "false",
    approved: "true",
    cad_url: "www.google.com",
    photo_url: "www.google.com",
    id: 1,
    comments: [],
    updated_at: "2018-11-18T14:53:08.533957"
  }
}
const store = mockStore(initialState);

describe('<NewSubmission />', () => {
  describe('render()', () => {
    test('renders the component', () => {
      const wrapper = shallow(<NewSubmission store={store} />);
      // const component = wrapper.dive();

      expect(toJson(wrapper)).toMatchSnapshot();
    });

    // test to render model from contentful
  });
});
