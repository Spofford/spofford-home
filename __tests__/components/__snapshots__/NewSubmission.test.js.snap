// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewSubmission /> render() renders the component 1`] = `
<NewSubmission
  dispatch={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  submission={
    Object {
      "advance": "false",
      "approved": "true",
      "cad_url": "www.google.com",
      "comments": Array [],
      "description": "description one",
      "id": 1,
      "manufacturing": "manufacturing description one",
      "photo_url": "www.google.com",
      "updated_at": "2018-11-18T14:53:08.533957",
    }
  }
  user={
    Object {
      "charges": Array [],
      "email": "amhasler@gmail.com",
      "first_name": "Adam",
      "id": 20,
      "last_name": "Hasler",
      "role": 1,
    }
  }
/>
`;
